### PR TITLE
Fix to add parentheses for TSTypeQuery in some case

### PR DESCRIPTION
**Pull request by onishi-kohei**
## Description
Fix to add parentheses for TSTypeQuery in some case (#13546)

I added an if statement with the parent type the example sentence had.
``` js
a as (typeof node.children)[number]

a as (typeof node.children)[]

a as ((typeof node.children)[number])[]
````

## Checklist

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
Need add tests https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md#tests

File: tests/format/typescript/typeof/typeof.ts [hunk 1]
```diff
@@ -0,0 +1,4 @@
+// https://github.com/typescript-eslint/typescript-eslint/pull/4382
```
Comment (thorn0) [inline tests/format/typescript/typeof/typeof.ts:1]: Please remove this comment

**Comment by thorn0** [comment]
Please add a changelog entry. See https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md#other

**Comment by onishi-kohei** [comment]
I am grateful to @thorn0  and @fisker for their support.

File: src/language-js/needs-parens.js [hunk 2]
```diff
@@ -456,7 +456,14 @@ function needsParens(path, options) {
...
       );
-
+    case "TSTypeQuery":
+      if (
+        parent.type === "TSIndexedAccessType" ||
+        parent.type === "TSArrayType"
+      ) {
+        return true;
+      }
+    // fallthrough
```
Comment (fisker) [inline src/language-js/needs-parens.js:466]: I don't think `TSTypeQuery`(TypeScript node type) can be a child of `NullableTypeAnnotation`(flow node type). This should not fallthrough.

**Review by fisker — CHANGES_REQUESTED** [review]

File: src/language-js/needs-parens.js [hunk 3]
```diff
@@ -456,7 +456,14 @@ function needsParens(path, options) {
...
           path.getParentNode(1).type.startsWith("TSJSDoc"))
       );
-
+    case "TSTypeQuery":
+      if (
+        parent.type === "TSIndexedAccessType" ||
+        parent.type === "TSArrayType"
+      ) {
+        return true;
+      }
```
[the remark this answers is not in the document]
Comment (onishi-kohei) [reply src/language-js/needs-parens.js:465]:
```js
a as (typeof node.children)[]
```
this property name was `elementType`
May I add a case like the code below?

’’’ js
return (
        name === "objectType" || name === "elementType"   &&
        (parent.type === "TSIndexedAccessType" || parent.type === "TSArrayType")
      )
'''

[the remark this answers is not in the document]
Comment (fisker) [reply src/language-js/needs-parens.js:465]: No, should be 

```js
return (
        (name === "objectType" && parent.type === "TSArrayType")&&
        (name === "elementType" && parent.type === "TSIndexedAccessType" )
      )
```

**Review by sosukesuzuki — COMMENTED** [review]
LGTM except for the review comments that remain now.

File: src/language-js/needs-parens.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (onishi-kohei) [reply src/language-js/needs-parens.js]: Fixed in this commit.　https://github.com/prettier/prettier/pull/14042/commits/70353171ed013c8e40b0a425d7d8b927ea690bbc

**Review by fisker — APPROVED** [review]
Thanks for working on this.

**Review by sosukesuzuki — APPROVED** [review]

**Comment by fisker** [comment]
@onishi-kohei  

There should be other cases need add parens, for example:

```js
a as (typeof node.children)!
```

if you are interested [thoses types in this branch](https://github.com/prettier/prettier/blob/369bf2a1c117bdf4346e851d1039d6d2b014e160/src/language-js/needs-parens.js#L500) may give you the idea to improve.

And don't send PRs from your `main` branch next time.

**Comment by SamuelMS** [comment]
@onishi-kohei Does this interact cleanly with `keyof typeof`? I'm seeing new errors in `2.8.3`:

```typescript
{
  // ...
  accept: typeof MIME_TYPES[keyof typeof MIME_TYPES][];
}
```

> ```14:11  Error: Replace `typeof·MIME_TYPES` with `(typeof·MIME_TYPES)`  prettier/prettier```

The type is ugly but is tied to a third-party library. Parentheses that are added by hand are stripped by prettier automatically.

**Comment by onishi-kohei** [comment]
@SamuelMS 
I have verified that the following code did not generate an error.
``` js
// Input
export type A = {
  a?: {
    [
      x: string
    ]: typeof SomeLongLongLongTypeName[keyof typeof SomeLongLongLongTypeName];
  } | null;
};
```

**Comment by thorn0** [comment]
@SamuelMS Make sure you don't have a caret (`^`) in front of Prettier's version in your `package.json`. See https://prettier.io/docs/en/install.html#summary

**Comment by zomars** [comment]
Can we disable this option?

**Comment by SamuelMS** [comment]
@onishi-kohei You're right, this was an environment issue – IDEA silently persisted a cached version of an older version of prettier around.

**Pull request merged**